### PR TITLE
Add unrolled functions and unit tests

### PR DESCRIPTION
**Pull request by dennisYatunin**
## Purpose 

This PR moves `ClimaCore.Utilities.UnrolledFunctions` into `UnrolledUtilities.jl`. The definitions of generated functions copied over from `Unrolled.jl` have been simplified and generalized to multiple input arguments, while the rest is mostly unchanged.

This also adds a thorough set of unit tests for the unrolled functions, which includes a comparison of performance and compilation between `UnrolledUtilities` and corresponding functions from `Base` and `Base.Iterators`. The comparison results are currently printed as a table in CI; I will move these into a docs page in a follow-up PR.

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

**Review by charleskawczynski — APPROVED** [review]

File: src/UnrolledUtilities.jl [hunk 1]
```diff
@@ -1,5 +1,164 @@
...
+  of the corresponding functions from `Base` encounter type-instabilities and
+  allocations when this is the case)
+- the iterators have many elements (e.g., more than 32, which is the threshold
+  at which `map` becomes type-unstable for `Tuple`s)
+- `f` and/or `op` recursively call the function to which they is passed, with an
+  arbitrarily large recursion depth (e.g., if `f` calls `map(f, itrs)`, it will
+  be type-unstable when the recursion depth exceeds 3, but this will not be the
+  case with `unrolled_map`)
+
+In addition, these functions are guaranteed to be optimized out through constant
```
Comment (charleskawczynski) [inline src/UnrolledUtilities.jl]: I think that this is maybe too strong of a statement, can we maybe reword to “A good deal of effort was made to optimized out through constant propagation…”?

Comment (dennisYatunin) [reply src/UnrolledUtilities.jl]: I added a bunch of `@inline`s just now, and it looks like everything that should be constant is now constant. But yeah, this is a pretty strong statement; I'll change "guaranteed" to "very likely".

Comment (charleskawczynski) [reply src/UnrolledUtilities.jl]: > I think that this is maybe too strong of a statement, can we maybe reword to “A good deal of effort was made to optimized out through constant propagation…”?
